src/Importer.lua

```diff
@@ -105,29 +166,58 @@ function Importer:import(callingScript, path, exports)
 	assert(importCheck(callingScript, path, exports))
 
 	local parts = path:split("/")
-	local current =  callingScript
+	local current = callingScript
 	local hasAscendedParents = false
+	local isFirstPart = true
 
 	for _, pathPart in pairs(parts) do
-		local nextInstance = self:getNextInstance(current, pathPart, hasAscendedParents)
+		local nextInstance = self:getNextInstance(current, pathPart, hasAscendedParents, isFirstPart)
 
+		if isFirstPart then
+			assert(nextInstance, ("'%s' is not the name of a service, alias, or child of current dataModel (%s)")
+				:format(pathPart, self.dataModel.Name))
+		else
+			assert(nextInstance, ("Could not find a child '%s' at \"%s.%s\"")
+				:format(pathPart, current:GetFullName(), pathPart))
+		end
 		-- This makes sure that `../` will take you up into the parent of the
 		-- script (script.Parent.Parent), but `../../` will only take you up
 		-- one extra parent after that.
 		if pathPart == ".." then
 			hasAscendedParents = true
 		end
 
-		assert(nextInstance, ("Could not find \"%s.%s\""):format(current:GetFullName(), pathPart))
-
 		current = nextInstance
+		isFirstPart = false
 	end
 
 	if current:IsA("ModuleScript") then
+		local module
+
+		if self._config.detectRequireLoops then
+			local isRecursivelyRequiring, startIndex = self:checkIfAlreadyWaitingForRequire(current)
+			if isRecursivelyRequiring then
+				local recursionPathStr = current.Name
+				for i = startIndex+1, #self._currentlyRequiring do
+					local module = self._currentlyRequiring[i]
+					recursionPathStr = recursionPathStr .. " - > " .. module.Name
+				end
+				recursionPathStr = recursionPathStr .. " - > " .. current.Name
+
+				error(("Require loop! %s"):format(recursionPathStr))
+			end
+
+			self:setIsCurrentlyRequiring(current, true)
+			module = require(current)
+			self:setIsCurrentlyRequiring(current, false)
+		else
+			module = require(current)
+		end
+
```

Comment:
Could you explain what this is doing? And also, could we break this into a new function with a descriptive name?

Reply:
So basically, any time we call require() on a module, all the code inside of it will run before it returns its value. Because of this, if we have a bunch of modules that import other modules in a chain, each one will get added to a list as currently being required (line 210) until that recursive chain completes, at which point the module is removed from the list.

Before we try to require the module, we iterate over that table of modules we're currently requiring, to test to see if its already being required further up the chain (line 198)

The function we call on 198 also returns an index, which is the position of the module that's trying to be loop-required. From that index, to the end of the list, represents the path which the module is being required in a loop, which we build out as a string from line 200 to 205.

Reply:
refactored this ^-^

Reply:
cool as hell, thanks for the explanation